logstash-core/src/main/java/org/logstash/log/PipelineRoutingAppender.java

```diff
@@ -0,0 +1,157 @@
+ *          <SizeBasedTriggeringPolicy size="100MB" />
+ *          <DefaultRolloverStrategy max="30" />
+ *        </RollingFile>
+ *  </PipelineRouting>
+ *  }
+ * </pre>
+ * */
+@Plugin(name = "PipelineRouting", category = Core.CATEGORY_NAME, elementType = Appender.ELEMENT_TYPE, printObject = true, deferChildren = true)
+public class PipelineRoutingAppender extends AbstractAppender {
+
+    /**
+     * Builder for {@link PipelineRoutingAppender} instances
+     * */
+    public static class Builder<B extends PipelineRoutingAppender.Builder<B>> extends AbstractAppender.Builder<B>
+            implements org.apache.logging.log4j.core.util.Builder<PipelineRoutingAppender> {
+
+        @PluginNode
+        private Node appenderNode;
+
+        @Override
+        public PipelineRoutingAppender build() {
+            final String name = getName();
+            if (name == null) {
+                LOGGER.error("No name defined for this RoutingAppender");
+                return null;
+            }
+            return new PipelineRoutingAppender(name, appenderNode, getConfiguration());
+        }
+    }
+
+    /**
+     * Factory method to instantiate the appender
+     * */
+    @PluginBuilderFactory
+    public static <B extends PipelineRoutingAppender.Builder<B>> B newBuilder() {
+        return new PipelineRoutingAppender.Builder<B>().asBuilder();
+    }
+
+    private final Node appenderNode;
+    private final Configuration configuration;
+    private final ConcurrentMap<String, AppenderControl> createdAppenders = new ConcurrentHashMap<>();
+    private final Map<String, AppenderControl> createdAppendersUnmodifiableView =
+            Collections.unmodifiableMap(createdAppenders);
+
+    protected PipelineRoutingAppender(String name, Node appenderNode, Configuration configuration) {
+        super(name, null, null, false, new Property[0]);
+        this.appenderNode = appenderNode;
+        this.configuration = configuration;
+    }
+
+    /**
+     * Returns an unmodifiable view of the appenders created by this {@link PipelineRoutingAppender}.
+     */
+    public Map<String, AppenderControl> getAppenders() {
+        return createdAppendersUnmodifiableView;
+    }
+
+    /**
+     * Core method to apply the logic of routing.
+     * */
+    @Override
+    public void append(LogEvent event) {
+        AppenderControl appenderControl = getControl(event);
+
+        if (appenderControl != null) {
+            appenderControl.callAppender(event);
+        }
+    }
+
+    /**
+     * Create or retrieve the sub appender for the pipeline.id provided into the event
+     * */
+    private AppenderControl getControl(LogEvent event) {
+        final String key;
+        if (event.getContextData().containsKey("pipeline.id")) {
+            key = event.getContextData().getValue("pipeline.id");
+        } else {
+            error("Unable to find the pipeline.id in event's context data");
+            key = "sink";
+        }
```

Comment:
nit: intent here seems like `"pipeline.id"` in the map would be optional
```suggestion
    String key = event.getContextData().getValue("pipeline.id");
    if (key == null) {
        error("Unable to find the pipeline.id in event's context data");
        key = "sink";
    }
```
[docs](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/util/ReadOnlyStringMap.html#getValue-java.lang.String-) is clear about returning null or missing mapping (just like a plain-old Java map)